### DISCUSSION TITLE
Typescript Smart Component examples

### DISCUSSION THREAD

**Discussion by mortocks** [react-hook-form/react-hook-form#9392] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

Cannot find good examples of how to use Typescript with Smart Components

**Describe the solution you'd like**

On the [following documentation page](https://react-hook-form.com/advanced-usage#SmartFormComponent) please provide an example of how to create a basic "Smart Component" with typescript.

**Describe alternatives you've considered**

Lots of googling 🤷‍♂️ reading through the [react hook form types documentation](https://react-hook-form.com/ts/) and there are examples for v6 and for non typescript components but nothing that specifies how to correctly type a smart component 

**Additional context**

Here's the example from the docs about how to create a smart component, but can't find any solutions to bringing in [typescript support](https://react-hook-form.com/ts/)

```js
export function Input({ register, name, ...rest }) {
  return <input {...register(name)} {...rest} />;
}
```

Below is some sudo code I've tried. It compiles correctly but as soon as you type anything in the field, the property "name" becomes `undefined`.

```ts
interface Props extends InputHTMLAttributes<HTMLInputElement> {
  foo: string;
}

const SmartInput = ({ foo, ...rest }: Props) => {
  return (
    <div>
      {foo}
      <input {...rest} />
    </div>
  );
}

<form ...>
  const { register } = useForm(...)
...
  <SmartInput
    foo="bah"
    {...register("name")}
    ...
  />
</form>

```

Note, there is [this example](https://codesandbox.io/s/react-hook-form-useformmethods-e3411?file=/src/App.tsx:67-83) in the "examples" folder but it requires additionally wrapping the component in a custom form component. Trying to create something like the simple JS example

**Comment by jacobpenny** [comment, marked as answer]
I think your code isn't working because the `ref` that `register` returns isn't getting passed through to the `input`. 

This is working for me:
```typescript
interface Props extends InputHTMLAttributes<HTMLInputElement> {
  foo: string;
}

export const SmartInput = React.forwardRef<HTMLInputElement, Props>(
  ({ foo, ...rest }: Props, ref) => {
    return (
      <div>
        {foo}
        <input ref={ref} {...rest} />
      </div>
    );
  }
);
```

**Reply by eliawk** [reply]
@jacobpenny I've tried to implement your solution but I still can't get the values updated from the input fields and get this error in `SmartInput`: **'ref' is specified more than once, so this usage will be overwritten**

**Comment by mortocks** [comment]
Thanks @jacobpenny - solution worked perfectly. Just had to add `SmartInput.displayName = "SmartInput";` to keep the linter happy. Thanks 🥳

**Comment by AndresdoSantos** [comment]
Thanks @jacobpenny
